In [37]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

In [38]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df=pd.read_csv(Path("../Supervised-Learning_Module12/lending_data.csv"))

# Review the DataFrame
lending_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

In [39]:
# Separate the data into labels and features

# Separate the y variable, the labels
y=lending_df['loan_status']

# Separate the X variable, the features
X=lending_df.drop(columns=['loan_status']).copy()

In [40]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [41]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [42]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

In [43]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train,X_test,y_train,y_test=train_test_split(X, y, random_state=1)

In [47]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train,y_train)

In [48]:
# Make a prediction using the testing data
training_predictions=lr_model.predict(X_train)
testing_predictions=logistic_regression_model.predict(X_test)

In [53]:
# Print the balanced_accuracy score of the model
balanced_accuracy_score(y_test, testing_predictions)

0.9520479254722232

In [59]:
# Generate a confusion matrix for the model
testing_matrix = confusion_matrix(y_test, testing_predictions)
print(testing_matrix)

[[18663   102]
 [   56   563]]


In [61]:
# Print the classification report for the model
testing_report=classification_report_imbalanced(y_test,testing_predictions)
print(testing_report)

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.91      1.00      0.95      0.91     18765
          1       0.85      0.91      0.99      0.88      0.95      0.90       619

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



The model predicts the healthy loans nearly perfect and the high-risk loans with 85% precision and 91% recall.

In [62]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros=RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_oversampled, y_oversampled = ros.fit_resample(X_train,y_train)

In [64]:
# Count the distinct values of the resampled labels data
y_oversampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

In [65]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
oversampled_lr_model = logistic_regression_model.fit(X_oversampled,y_oversampled)

# Make a prediction using the testing data
os_training_predictions=lr_model.predict(X_oversampled)
os_testing_predictions=logistic_regression_model.predict(X_test)

In [66]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(y_test, os_testing_predictions)

0.9936781215845847

In [67]:
# Generate a confusion matrix for the model
os_testing_matrix = confusion_matrix(y_test, os_testing_predictions)
print(os_testing_matrix)

[[18649   116]
 [    4   615]]


In [68]:
# Print the classification report for the model
os_testing_report=classification_report_imbalanced(y_test,os_testing_predictions)
print(os_testing_report)

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18765
          1       0.84      0.99      0.99      0.91      0.99      0.99       619

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



The logistic regression model, fit with oversampled data, predicts  the healthy loans nearly perfect and the high-risk loans with 84% precision and 99% recall.